include/ginkgo/core/matrix/fft.hpp

```diff
@@ -0,0 +1,308 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_MATRIX_FFT_HPP_
+#define GKO_PUBLIC_CORE_MATRIX_FFT_HPP_
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+
+
+namespace gko {
+namespace matrix {
+
+
+/**
+ * This LinOp implements a 1D Fourier matrix using the FFT algorithm.
+ *
+ * It implements real-to-complex and complex-to-complex forward and inverse DFT.
+ *
+ * For a power-of-two size n with corresponding root of unity
+ * $\omega = e^{-2\pi i / n}$ for forward DFT and $\omega = e^{2 \pi i / n}$
+ * for inverse DFT it computes
+ *
+ *     $$ x_k = \sum_{i=0}^{n-1} \omega^{ik} b_i $$
+ *
+ * without normalization factors.
+ *
+ * @ingroup identity
+ * @ingroup mat_formats
+ * @ingroup LinOp
+ */
+class Fft : public EnableLinOp<Fft>,
```

Comment:
Is there a chance we need more than 4D FFT? Personally, I would prefer if we had a template parameter for the dimensionality:
```
template <int dimensionality>
class Fft : public EnableLinOp<Fft>,
```

Reply:
I would prefer having individual types, since otherwise we would need variadic or gko::dim<dimensionality> parameters, which would complicate the interface, and since they don't share any implementation details except for transposition, this doesn't feel worth the reduced code duplication.

Reply:
OK that is fair.